### DISCUSSION TITLE
Colorized output from qt_sinks

### DISCUSSION THREAD

**Discussion by AnloWho** [gabime/spdlog#2277] [Q&A] [answered]
I was wondering if there's an easy way to get colorized output from the qt_sinks when working with QTextEdit? I have the sink working so I get log messages into the Qt application, but the coloring is obviously not there. Any tricks and tips?

**Comment by tt4g** [comment, marked as answer]
`spdlog::sinks::qt_sink` passes log messages to `QObject`, but does nothing more.

Check for the appearance of certain patterns in the log messages, and change the colors yourself.

**Reply by AnloWho** [reply]
Thanks I got it to work. Just created my own sink and intercepted the messages there for colorization. Pretty neat!

**Reply by qwertae** [reply]
Hi, could you please share the code (i am sure lots of other people are looking for the same). Regards

**Reply by AnloWho** [reply]
Ok, 2nd rewrite since my browser decided to shutdown on me before I finished the reply. Anyway, here goes:
You basically follow the example from the Wiki on creating a [custom sink](https://github.com/gabime/spdlog/wiki/4.-Sinks#implementing-your-own-sink). 

```
template <typename Mutex>
class QtLogSink: public spdlog::sinks::base_sink<Mutex>
{
public:
    QtLogSink(QTextEdit* textEdit)
    {
        textEdit_ = textEdit;
    }

    ~QtLogSink()
    {
        flush_();
    }

protected:
    void sink_it_(const spdlog::details::log_msg& message) override
    {
        spdlog::memory_buf_t formatted;
        spdlog::sinks::base_sink<Mutex>::formatter_->format(message, formatted);
        spdlog::string_view_t formattedMessage = spdlog::string_view_t(formatted.data(), formatted.size());

        auto outputString = QString::fromUtf8(formattedMessage.data(), static_cast<int>(formattedMessage.size())).trimmed();

        textEdit_->setTextColor(getMessageColor(message.level));
        textEdit_->append(outputString);
    }

    void flush_() override
    {
    }
private:
    QColor getMessageColor(spdlog::level::level_enum logLevel)
    {
     // here goes your logic to intercept the message and return the color. check for [info], [warning], etc.
     }

private:
    QTextEdit* textEdit_;
};

using qtLogSink_mt = QtLogSink<std::mutex>;
using qtLogSink_st = QtLogSink<spdlog::details::null_mutex>;
```
Then you can use this sink, e.g. add it to the default logger somewhere appropriate (where you have access to QTextEdit so you can use it when you create the qtLogSink_mt object for example.  Hope this helps.

**Reply by gabime** [reply]
Cool. The message level can be checked by accessing  `message.level` so no need to search for it in the message string.

**Reply by AnloWho** [reply]
Ok thanks. will consider using that then.

**Reply by AnloWho** [reply]
> Cool. The message level can be checked by accessing `message.level` so no need to search for it in the message string.

I update the example to use message.level. Works fine.
